File: pagerank.scala
import java.io._
import org.apache.spark._

  case class Row(url : String, pagerank: Double, urls: Array[String])

  val rawData = sc.textFile("datas/etl-xaaCC-MAIN.txt")

  var lines = rawData.map{ s =>
    val parts = s.split("\t")
    var links = parts(2).substring(1,parts(2).length -1).split(",")
    Row(parts(0), parts(1).toInt, links)
  }

  val contribs = lines.flatMap{ case Row(url,pagerank,urls) =>
    val size = urls.size
    urls.map(url => (url, pagerank / size))
  }

  val newPagerank = contribs.reduceByKey((x,y) => x+y).mapValues(v => 0.15 + 0.85*v)
  newPagerank.sortBy(-_._2).saveAsTextFile("result")

File: pagerank.py
#!/usr/bin/python 
from org.apache.pig.scripting import *
P = Pig.compile("""
previous_pagerank = LOAD '$docs_in' USING PigStorage('\t') AS ( url: chararray, pagerank: float, links:{ link: ( url: chararray ) } );
outbound_pagerank = FOREACH previous_pagerank GENERATE pagerank / COUNT ( links ) AS pagerank, FLATTEN ( links ) AS to_url; 
new_pagerank = FOREACH ( COGROUP outbound_pagerank BY to_url, previous_pagerank BY url INNER ) GENERATE group AS url, ( 1 - $d ) + $d * SUM ( outbound_pagerank.pagerank ) AS pagerank, FLATTEN ( previous_pagerank.links ) AS links;
ordered = ORDER new_pagerank BY pagerank DESC;
STORE ordered INTO '$docs_out' USING PigStorage('\t');""")

# 'd' tangling value in pagerank model
params = { 'd': '0.85', 'docs_in': "datas/etl-xaaCC-MAIN.txt" }

for i in range(5):
	output = "output/pagerank_data_" + str(i + 1)
	params["docs_out"] = output
	stats = P.bind(params).runSingle()
	if not stats.isSuccessful():
		raise 'failed'
	params["docs_in"] = output
 
